5.1 regression from 5.0: JavaScript heap out of memory (Type instantiation is excessively deep and possibly infinit)

# Bug Report

`5.0.4` used to work fine after upgrading to the current latest version `5.1.3`. Very slow builds and eventually crashes. My root-cause seems to be a circular issue of some sort with tsc.

Seems like tsc can't handle interface definition files anymore, in this case from `lodash`?

```
node_modules/@types/lodash/common/common.d.ts:194:15 - error TS2589: Type instantiation is excessively deep and possibly infinite.

194     interface Object<T> extends LoDashImplicitWrapper<T> {
                  ~~~~~~

node_modules/@types/lodash/common/common.d.ts:206:15 - error TS2430: Interface 'ObjectChain<T>' incorrectly extends interface 'LoDashExplicitWrapper<T>'.
  The types returned by 'entries().pop()' are incompatible between these types.
    Type 'CollectionChain<string | T[keyof T]>' is missing the following properties from type 'ObjectChain<[string, any]>': assign, assignIn, assignInWith, assignWith, and 11 more.

206     interface ObjectChain<T> extends LoDashExplicitWrapper<T> {
                  ~~~~~~~~~~~
```

My workaround was adding `"skipLibCheck": true` to the tsconfig.json did mitigate the issue for now.


```
node_modules/@types/lodash/common/common.d.ts:194:15 - error TS2589: Type instantiation is excessively deep and possibly infinite.

194     interface Object<T> extends LoDashImplicitWrapper<T> {
                  ~~~~~~

node_modules/@types/lodash/common/common.d.ts:206:15 - error TS2430: Interface 'ObjectChain<T>' incorrectly extends interface 'LoDashExplicitWrapper<T>'.
  The types returned by 'entries().pop()' are incompatible between these types.
    Type 'CollectionChain<string | T[keyof T]>' is missing the following properties from type 'ObjectChain<[string, any]>': assign, assignIn, assignInWith, assignWith, and 11 more.

206     interface ObjectChain<T> extends LoDashExplicitWrapper<T> {
                  ~~~~~~~~~~~
```
Workaround for me was to add: `"skipLibCheck": true` to the tsconfig.json did mitigate the issue for now.


### 🔎 Search Terms

- Crash
- OOME
- 5.1
- Heap

### 🕗 Version & Regression Information

I also tried the latest `next` tag, with the same issues.

- This is a crash
- This changed between versions 5.0.x and 5.1.x

### ⏯ Playground Link

The "Type is excessively deep" is being triggered in a project that includes this the `@types/lodash`, so in your package.json be sure you have the dependency:

```json
"dependencies": {
     "lodash": "^4.17.21"
},
"devDependencies": {
    "@types/lodash": "^4.14.195"   
}
```

Be sure that you did NOT set `skipLibCheck` to `true`. So keep it to false, otherwise the `d.ts` files will be ignored. 

Execute via: `NODE_OPTIONS=--max-old-space-size=8192 tsc --pretty --diagnostics --extendedDiagnostics --incremental false`

### 💻 Code

See below, it doesn't involve my code.

### 🙁 Actual behavior

JavaScript heap out of memory, because of circular/ too deep dependency?

```bash
node_modules/@types/lodash/common/common.d.ts:194:15 - error TS2589: Type instantiation is excessively deep and possibly infinite.

194     interface Object<T> extends LoDashImplicitWrapper<T> {
                  ~~~~~~

node_modules/@types/lodash/common/common.d.ts:206:15 - error TS2430: Interface 'ObjectChain<T>' incorrectly extends interface 'LoDashExplicitWrapper<T>'.
  The types returned by 'entries().pop()' are incompatible between these types.
    Type 'CollectionChain<string | T[keyof T]>' is missing the following properties from type 'ObjectChain<[string, any]>': assign, assignIn, assignInWith, assignWith, and 11 more.

206     interface ObjectChain<T> extends LoDashExplicitWrapper<T> {
                  ~~~~~~~~~~~
Found 2 errors in 1 file.

Errors  Files
     2  node_modules/@types/lodash/common/common.d.ts:194
Files:                          253
Lines of Library:              9875
Lines of Definitions:         73521
Lines of TypeScript:           3999
Lines of JavaScript:              0
Lines of JSON:                    0
Lines of Other:                   0
Identifiers:                 117524
Symbols:                    4856905
Types:                      1093043
Instantiations:            11883895
Memory used:               2379915K
Assignability cache size:    230447
Identity cache size:            772
Subtype cache size:             398
Strict subtype cache size:       10
I/O Read time:                0.15s
Parse time:                   1.17s
ResolveModule time:           0.21s
ResolveTypeReference time:    0.03s
ResolveLibrary time:          0.02s
Program time:                 1.70s
Bind time:                    0.47s
Check time:                 239.51s
transformTime time:           0.05s
Source Map time:              0.02s
commentTime time:             0.04s
I/O Write time:               0.02s
printTime time:               0.37s
Emit time:                    0.37s
Total time:                 242.05s
```

### 🙂 Expected behavior

No crashes or errors during `tsc`!

Same problem with lodash here. Any news about this resolution ? 

This is my error : 

```
 ../../node_modules/@types/lodash/common/common.d.ts(194,15): error TS2589: Type instantiation is excessively deep and possibly infinite.
       ../../node_modules/@types/lodash/common/common.d.ts(206,15): error TS2430: Interface 'ObjectChain<T>' incorrectly extends interface 'LoDashExplicitWrapper<T>'.
         The types returned by 'entries().pop()' are incompatible between these types.
           Type 'CollectionChain<string | T[keyof T]>' is missing the following properties from type 'ObjectChain<[string, any]>': assign, assignIn, assignInWith, assignWith, and 11 more.
```

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -21538,11 +21538,6 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     // Relate components directly before falling back to constraint relationships
                     // A type S[K] is related to a type T[J] if S is related to T and K is related to J.
                     if (result = isRelatedTo((source as IndexedAccessType).objectType, (target as IndexedAccessType).objectType, RecursionFlags.Both, reportErrors)) {
-                        // This does _not_ generalize - specific instantiations of `S[K]` and `T[J]` may be related, even if the indexed accesses generally are not.
-                        // For example, `S = {x: string, a: string}`, `T = {x: string, b: string}`, `K = J = "x"`. `S` and `T` are unrelated, but the result of executing
-                        // `S["x"]` and `T["x"]` _are_. Given that, we have to flag the object type comparison here as "unreliable", since while the generic result can reliably
-                        // be used in the affirmative case, it failing is not an indicator that the structural result will not succeed.
-                        instantiateType((source as IndexedAccessType).objectType, reportUnreliableMapper);
                         result &= isRelatedTo((source as IndexedAccessType).indexType, (target as IndexedAccessType).indexType, RecursionFlags.Both, reportErrors);
                     }
                     if (result) {

--- a/tests/baselines/reference/genericIndexedAccessVarianceComparisonResultCorrect.errors.txt
+++ b/tests/baselines/reference/genericIndexedAccessVarianceComparisonResultCorrect.errors.txt
@@ -1,0 +1,35 @@
+genericIndexedAccessVarianceComparisonResultCorrect.ts(25,1): error TS2322: Type 'T<A>' is not assignable to type 'T<B>'.
+  Property 'z' is missing in type 'A' but required in type 'B'.
+
+
+==== genericIndexedAccessVarianceComparisonResultCorrect.ts (1 errors) ====
+    class A {
+        x: string = 'A';
+        y: number = 0;
+    }
+    
+    class B {
+        x: string = 'B';
+        z: boolean = true;
+    }
+    
+    type T<X extends { x: any }> = Pick<X, 'x'>;
+    
+    type C = T<A>;
+    type D = T<B>;
+    
+    type C_extends_D = C extends D ? true : false;                                  // true
+    type PickA_extends_PickB = Pick<A, 'x'> extends Pick<B, 'x'> ? true : false;    // true
+    type TA_extends_TB = T<A> extends T<B> ? true : false;                          // should be true
+    
+    declare let a: T<A>;
+    declare let b: T<B>;
+    declare let c: C;
+    declare let d: D;
+    
+    b = a;      // should be no error
+    ~
+!!! error TS2322: Type 'T<A>' is not assignable to type 'T<B>'.
+!!! error TS2322:   Property 'z' is missing in type 'A' but required in type 'B'.
+!!! related TS2728 genericIndexedAccessVarianceComparisonResultCorrect.ts:8:5: 'z' is declared here.
+    c = d;

--- a/tests/baselines/reference/genericIndexedAccessVarianceComparisonResultCorrect.types
+++ b/tests/baselines/reference/genericIndexedAccessVarianceComparisonResultCorrect.types
@@ -46,7 +46,7 @@ type PickA_extends_PickB = Pick<A, 'x'> extends Pick<B, 'x'> ? true : false;    
 >false : false
 
 type TA_extends_TB = T<A> extends T<B> ? true : false;                          // should be true
->TA_extends_TB : true
+>TA_extends_TB : false
 >true : true
 >false : false
 

